### PR TITLE
Firewall

### PR DESCRIPTION
**Pull request by smashwilson**
Configure iptables to only accept API traffic from the CLB nodes.

### CODE REVIEW

File: provision.yml [hunk 1]
```diff
@@ -73,14 +73,6 @@
...
       etcd_cluster_token: "{{ lookup('password', '.etcd_cluster_token length=20') }}"
 
-- name: pause to ensure SSH is up
-  hosts: deconst-all
-  gather_facts: no
-  tasks:
-
-  - name: ensure SSH is listening on all hosts
-    local_action: wait_for host={{ hostvars[inventory_hostname].ansible_ssh_host }} port=22 state=started
-
```
Comment (smashwilson) [inline provision.yml:83]: This breaks on VPN.

Comment (rgbkrk) [inline provision.yml:83]: Ug.

**Comment by smashwilson** [comment]
Things that now work:
- SSH connections are restricted to whitelisted hosts.
- Exposed service ports are restricted to the load balancer.
- etcd can peer with itself through the cluster.

The remaining work now is to ensure that the containers that need to reach the local etcd service (etcd-watcher and the webhook service) are able to do so.

**Comment by rgbkrk** [comment]
How's a reboot with this?

**Comment by smashwilson** [comment]
Right, I'll need to test that too. The iptables-restore systemd service should reapply these rules as the network is brought back up, but it would be good to verify.

File: roles/firewalled/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  notify: restart docker service
+  tags: network
+
+- name: iptables rules
+  template: src=rules-save.j2 dest=/var/lib/iptables/rules-save
+  sudo: yes
+  tags: network
+
+- name: invoke and enable the iptables service
+  service: name=iptables-restore.service enabled=yes state=started
```
Comment (rgbkrk) [inline roles/firewalled/tasks/main.yml]: Perhaps I don't know how the service directive works, but I don't see where the service is declared.

Comment (smashwilson) [inline roles/firewalled/tasks/main.yml]: It's built in to CoreOS, but not enabled by default.

Comment (smashwilson) [inline roles/firewalled/tasks/main.yml]:
```
core@deconst-worker-001 ~ $ systemctl cat iptables-restore.service
# /usr/lib64/systemd/system/iptables-restore.service
[Unit]
Description=Restore iptables firewall rules
# if both are queued for some reason, don't store before restoring :)
Before=iptables-store.service
# sounds reasonable to have firewall up before any of the services go up
Before=network.target
Conflicts=shutdown.target

[Service]
Type=oneshot
ExecStart=/sbin/iptables-restore /var/lib/iptables/rules-save

[Install]
WantedBy=basic.target
```

**Comment by smashwilson** [comment]
Also, good call, the rules that Docker adds to the `DOCKER` chain on startup are short-circuiting my logic there on reboots. Back to the drawing board.

**Comment by smashwilson** [comment]
I've been going in circles on this all day. Restricting SSH access is trivial, but not the part that we should actually care about. Only permitting external network access to containers from the load balancer on ServiceNet is killing me.

According to the [docker advanced networking guide](https://docs.docker.com/articles/networking/#communication-between-containers-and-the-wider-world), adding an iptables rule at the beginning of the `DOCKER` chain should work to restrict access to a host's containers to only a single IP:

```
iptables -I DOCKER -i ext_if ! -s 8.8.8.8 -j DROP
```

But, weirdly, even if I just add a rule to defeat _incoming_ public traffic from eth0:

```
-A DOCKER -i eth0 -j DROP
```

It defeats **all** traffic that's _outbound_ from the containers, like "DNS lookups" or "git clones":

```
core@deconst-worker-001 ~ $ docker logs content-service-1

> content-service@1.0.0 start /usr/src/app
> node app.js

/usr/src/app/src/connection.js:99
    if (err) throw err;
                   ^
Error: getaddrinfo ENOTFOUND identity.api.rackspacecloud.com
    at errnoException (dns.js:44:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:94:26)
```

I didn't think that outbound traffic would even touch the `DOCKER` chain in the filter table. This has something to do with NAT, I assume?

@rgbkrk, any ideas for something I could look at tomorrow? You know this stuff a lot better than I do.

What I do have sort of working is a role that installs a set of fixed pre-docker firewall rules, kicking the docker daemon service if necessary, to give consistent results among the cases of "applying firewall rules to a running host" and "restoring firewall rules after a reboot". The docker daemon installs a bunch of iptables rules on boot, and creating a base set of rules that will behave consistently once those rules are overlaid is a bit of a challenge. You _can_ keep the docker daemon from touching iptables at all with a command-line flag, but then Ansible would need to discover each container's internal IP somehow, and I don't think we're registering enough information in the docker module to get that working.

**Comment by smashwilson** [comment]
So the thing that was stopping me up yesterday was that the response packets from established connections were being defeated. Whoops.

Once I worked that out, the only remaining issue was a missing `INPUT` rule to accept etcd traffic that originated from the docker bridge, so that containers can connect to etcd on the same host. Now that that's in place, we should be good to :shipit: :sparkles:

**Comment by rgbkrk** [comment]
Ah, neat. What a pain.

Props to you Ash.

**Pull request merged**